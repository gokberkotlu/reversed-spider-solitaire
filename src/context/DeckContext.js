import React, { useState, createContext, useEffect } from "react";
import { cardSeries } from '../utils/CardSeries';

export const DeckContext = createContext();

export const DeckProvider = props => {

    // States
    const [gameCards, setGameCards] = useState([[],[],[],[],[],[],[],[],[],[]]); // Represents cards on table
    const [remainingCards, setRemainingCards] = useState([]); // Cards outside the game
    const [numberOfCompleted, setNumberOfCompleted] = useState(Array(0)); // Shows number of completed series
    const [newCardPacks, setNewCardPacks] = useState(Array(5)); // Shows number of new card packs
    const [isGameStarted, setIsGameStarted] = useState(false); // Set state when any card action happened
    const [statesLog, setStatesLog] = useState([]); // Log States to Undo the Game

    // States: Score Datas
    const [score, setScore] = useState(500); // Game score which starts 500
    const [highScore, setHighScore] = useState(+localStorage.getItem('highScore')); // Recorded highscore to store in localStorage

    const dealTheCards = (cards, numberOfCards) => {
        
        // Save Changes When the New Card Pack Added
        logGameChanges();

        if(cards.length >= 10) {
            let gameSet = [...gameCards];
            for(let i = 0; i < numberOfCards; i++) {
                gameSet[i%10].push(cards[i]);
            }
            setGameCards(gameSet);
            setRemainingCards(cards.splice(numberOfCards));

            return gameSet;
        }
    }

    // arrangeReversedCards arranges the reversed card at the start
    const arrangeReversedCards = (gameSet) => {
        gameSet.forEach(cards => {
            for(let i = 0; i < cards.length - 1; i++) {
                cards[i].isCardReverse = true;
            }
        })
    }

    // suffleTheCards shuffles eight decks of cards
    const suffleTheCards = () => {
        let deskNumber = 8;
        let newArray = [];
        for(let i = 0; i < deskNumber; i++) {
            newArray = newArray.concat(
                [...cardSeries.map((card) => {
                    return {
                        "cardNum": card,
                        "cardDeck": (i + 1),
                        "isCardReverse": false
                    }
                })]);
        }
        var currentIndex = newArray.length,
            randomIndex;
        // While there remain elements to shuffle...
        while (0 !== currentIndex) {
            // Pick a remaining element...
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex--;
    
            // And swap it with the current element.
            [newArray[currentIndex], newArray[randomIndex]] = [
                newArray[randomIndex], newArray[currentIndex]
            ];
        }
        return newArray;
    }

    // logGameChanges takes snapshots before all card actions and stores them inside statesLog
    const logGameChanges = () => {
        let addToLog = {
            "gameCardsBefore": JSON.parse(JSON.stringify(gameCards)),
            "remainingCardsBefore": JSON.parse(JSON.stringify(remainingCards)),
            "numberOfCompletedBefore": JSON.parse(JSON.stringify(numberOfCompleted)),
            "newCardPacksBefore": JSON.parse(JSON.stringify(newCardPacks))
        }
        setStatesLog(prevState => [...prevState, addToLog]);
    }

    // undoTheGame sets game states with last logged snapshot
    const undoTheGame = () => {
        let statesLogValue = [...statesLog];
        let statesLogValueLength = statesLogValue.length
        if(statesLogValueLength > 1) {
            let beforeStates = statesLogValue[statesLogValueLength - 1]; // Go back in the game by having "before value of last screen"
            setGameCards(beforeStates.gameCardsBefore);
            setRemainingCards(beforeStates.remainingCardsBefore);
            setNumberOfCompleted(beforeStates.numberOfCompletedBefore);
            setNewCardPacks(beforeStates.newCardPacksBefore);

            // Delete current states from log when went back in the game
            setStatesLog(statesLogValue.slice(0, statesLogValueLength - 1));

            // Decrease 1 point from score for each undo click
            setScore(prevState => prevState - 1);

            // Decrease 100 points from score if completed desk is removed by clicking undo button
            if(beforeStates.numberOfCompletedBefore.length < numberOfCompleted.length) {
                let scoreDiff = numberOfCompleted.length - beforeStates.numberOfCompletedBefore.length;
                setScore(prevState => prevState - 100 * scoreDiff);
            }
        }
    }

    useEffect(() => {
        // Shuffle cards, arrange reversed cards and deal them on the game table at start
        var allCards = suffleTheCards();
        var gameSet = dealTheCards(allCards, 54);
        arrangeReversedCards(gameSet);
    }, [])

    return (
        <DeckContext.Provider
        value={{
                "gameCards": [gameCards, setGameCards],
                "remainingCards": [remainingCards, setRemainingCards],
                "numberOfCompleted": [numberOfCompleted, setNumberOfCompleted],
                "newCardPacks": [newCardPacks, setNewCardPacks],
                "isGameStarted": [isGameStarted, setIsGameStarted],
                "suffleTheCards": suffleTheCards,
                "dealTheCards": dealTheCards,
                "arrangeReversedCards": arrangeReversedCards,
                "statesLog": [statesLog, setStatesLog],
                "logGameChanges": logGameChanges,
                "undoTheGame": undoTheGame,
                "score": [score, setScore],
                "highScore": [highScore, setHighScore]
            }}>
            {props.children}
        </DeckContext.Provider>
    );
}