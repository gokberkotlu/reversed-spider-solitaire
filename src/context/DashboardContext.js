import React, { useState, createContext, useContext, useEffect } from "react";
import { DeckContext } from "./DeckContext";

export const DashboardContext = createContext();

export const DashboardProvider = props => {

    // States
    const [counter, setCounter] = useState("00:00:00"); // Represents timer
    const [isGameFinished, setIsGameFinished] = useState(false); // According to this state, popup is shown or not
    const [resetBit, setResetBit] = useState(false); // It triggers the useEffect to reset the game
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = useState(0); // Decrease 2 points from score for each 10 sec

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [, setGameCards] = deckContextValue.gameCards;
    const [, setRemainingCards] = deckContextValue.remainingCards;
    const [, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [, setNewCardPacks] = deckContextValue.newCardPacks;
    const [, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setStatesLog] = deckContextValue.statesLog;
    const suffleTheCards = deckContextValue.suffleTheCards;
    const dealTheCards = deckContextValue.dealTheCards;
    const arrangeReversedCards = deckContextValue.arrangeReversedCards;

    const [, setScore] = deckContextValue.score;

    const resetGame = () => {
        // Resetting
        setGameCards([[],[],[],[],[],[],[],[],[],[]]);
        setRemainingCards([]);
        setNewCardPacks(Array(5));
        setNumberOfCompleted(Array(0));
        setScore(500);
        setCounter("00:00:00");
        setIsGameFinished(false);
        setResetBit(true);
        setIsGameStarted(false);
        setStatesLog([]);
        setDecreaseInEveryTenSeconds(0);
    }

    useEffect(() => {// Deal cards again from scratch
        if(resetBit) {
            var allCards = suffleTheCards();
            var gameSet = dealTheCards(allCards, 54); // setGameCards and setRemainingCards included
            arrangeReversedCards(gameSet);
            setResetBit(false);
        }
    }, [resetBit]) // resetBit changes trigger the control of gameSet and remainingCards

    return (
        <DashboardContext.Provider
        value={{"counter": [counter, setCounter],
        "isGameFinished": [isGameFinished, setIsGameFinished],
        "resetGame": resetGame,
        "decreaseInEveryTenSeconds": [decreaseInEveryTenSeconds ,setDecreaseInEveryTenSeconds]}}>
            {props.children}
        </DashboardContext.Provider>
    );
}