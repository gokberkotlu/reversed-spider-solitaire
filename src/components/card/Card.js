import React, { useContext, useState } from 'react';
import CardBack from './card-templates/card-back/CardBack';
import CardFront from './card-templates/card-front/CardFront';
import { DeckContext } from '../../context/DeckContext';
import { cardSeries } from '../../utils/CardSeries';
import './card.css'

const Card = ({cardNumber, cardUnitNumber, cardDeck, isCardReverse}) => {

    const deckContextValue = useContext(DeckContext);
    const [gameCards,] = deckContextValue.gameCards;

    const [mousePositionX, setMousePositionX] = useState(0);
    const [mousePositionY, setMousePositionY] = useState(0);
    var xAxisDifference;
    var yAxisDifference;

    const dragStart = (event) => {

        if(event.target.className === 'card') {
            // Setdata parameter ************//
            let cardNumber = event.target.id.split("-")[0];
            let cardDeckNumber = +event.target.id.split("-")[4];
            let cardUnitIndex = event.target.id.split("-")[2] - 1;
            let cardsInTheUnit = gameCards[cardUnitIndex];

            // Find the index of card in current unit
            var unitIndexOfCard;
            
            gameCards[cardUnitIndex].forEach((card, index) => {
                if(card.cardNum === cardNumber && card.cardDeck === cardDeckNumber) {
                    unitIndexOfCard = index;
                }
            })

            // Set the data to be sent with dataTransfer.setData
            var sendData = [];
            var lastNumber;

            // Loop to get movable cards by comparing them each other until the end of unit
            for(let i = unitIndexOfCard; i < cardsInTheUnit.length; i++) {
                let cardNumValue = cardsInTheUnit[i].cardNum;
                if(sendData.length === 0) {
                    sendData.push(event.target.id);
                    lastNumber = cardNumValue;
                } else {
                    if(cardSeries.indexOf(lastNumber) === cardSeries.indexOf(cardNumValue) - 1 && !isCardReverse) {
                        sendData.push(`${cardsInTheUnit[i].cardNum}-spare-${cardUnitIndex + 1}-deck-${cardsInTheUnit[i].cardDeck}`);
                        lastNumber = cardNumValue;
                    } else {
                        sendData = [];
                        break;
                    }
                }
            }

            // Send if informations of movable cards
            event.dataTransfer.setData("Text", sendData);
            event.dataTransfer.setDragImage(new Image("0", "0"), -10, -10);

            // Add selected__card classname to movable cards
            for(let i = 0; i < sendData.length; i++) {
                document.getElementById(sendData[i]).children[0].classList.add("selected__card");
            }

            setMousePositionX(event.pageX);
            setMousePositionY(event.pageY);
        }
    }

    const drag = (event) => {
        const selectedCards = document.querySelectorAll(".selected__card");
        xAxisDifference = event.pageX - mousePositionX;
        yAxisDifference = event.pageY - mousePositionY;
        // Drag card elements by axis difference
        var css = "z-index:9999; pointer-events: none; transform: translate(" + xAxisDifference + "px, " +
        yAxisDifference + "px)";
        selectedCards.forEach(element => {
            element.style.cssText = css;
        })
    }

    const dropEnd = () => {
        // Bring the card elements to initial state when drop end
        var css = "z-index:0; pointer-events:auto; transform:translate(0,0);";
        document.querySelectorAll(".selected__card").forEach((element) => {
            element.classList.remove("selected__card");
            element.style.cssText = css;
        });
    }

    return (
        <div id={`${cardNumber}-spare-${cardUnitNumber}-deck-${cardDeck}`}
        className="card"
        onDragStart={(event) => dragStart(event)}
        onDrag={(event) => drag(event)}
        onDragEnd={() => dropEnd()}
        draggable>
            {isCardReverse && <CardBack/>}
            {!isCardReverse && <CardFront cardNumber={cardNumber} />}
        </div>
    );
}

export default React.memo(Card);