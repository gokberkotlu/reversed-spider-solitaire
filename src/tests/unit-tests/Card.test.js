import renderer from "react-test-renderer";
import Card from '../../components/card/Card';
import { DeckProvider } from '../../context/DeckContext';

describe('Card component', () => {
    it('should render Card component', () => {
        const tree = renderer.create(
            <DeckProvider value={null}>
                <Card />
            </DeckProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
})