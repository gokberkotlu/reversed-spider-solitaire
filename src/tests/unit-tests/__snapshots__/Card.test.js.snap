// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render Card component 1`] = `
<div
  className="card"
  draggable={true}
  id="undefined-spare-undefined-deck-undefined"
  onDrag={[Function]}
  onDragEnd={[Function]}
  onDragStart={[Function]}
>
  <div
    className="card-template card-spade card-front"
  >
    <div>
      <div
        className="card-rank"
      />
      <div
        className="spade-icon"
      >
        ♠
      </div>
    </div>
    <div
      className="spade-center"
    >
      ♠
    </div>
    <div>
      <div
        className="spade-icon reversed-icon"
      >
        ♠
      </div>
      <div
        className="card-rank reversed-icon"
      />
    </div>
  </div>
</div>
`;
